[RFC] citra-qt: Print stack trace

MerryMage opened the pull request:
This PR implements printing of a stack trace to the console when something bad happens. On Windows and OS X it also displays a modal message box to the user.

![20160602-135012](https://cloud.githubusercontent.com/assets/8682882/15754567/9312815a-28f0-11e6-946a-57385dca3175.png)
![](https://i.imgur.com/fNcRjcQ.png)
*Example of an `ASSERT_MSG` being caught.*

## Motivation

To hopefully get more useful reports from users rather than just "it crashes".

Could *possibly* be somehow extended to dump logs to a file in that case and/or integrate with an issue tracking system.

## Known isuses

* It's in `citra-qt` because I was originally going to use `QMessageBox`, but that doesn't work at all. Perhaps I should move it to `common`.
* Untested on ~~OS X and~~  Linux at the moment.
* Consider [localisation](https://github.com/citra-emu/citra/pull/1874#discussion_r65587189) of user-facing strings.

lioncash commented on src/citra_qt/crash_handler.cpp:
```diff
@@ -0,0 +1,270 @@
+    };
+
+    // Initialise symbols
+    if (SymInitialize(process, NULL, TRUE) == FALSE) {
```
nullptr

lioncash commented on src/citra_qt/crash_handler.cpp:
```diff
@@ -0,0 +1,270 @@
+
+    // Walk the stack
+    do {
+        if (!StackWalk64(IMAGE_FILE_MACHINE_AMD64, process, thread, &s, &c, NULL, SymFunctionTableAccess64, SymGetModuleBase64, NULL)) {
```
nullptr

lioncash commented on src/citra_qt/crash_handler.cpp:
```diff
@@ -0,0 +1,270 @@
+    OSXMessageBox(title, message);
+#endif
+
+    exit(-1);
```
```cpp
exit(EXIT_FAILURE);
```

lioncash commented on src/citra_qt/crash_handler.cpp:
```diff
@@ -0,0 +1,270 @@
+}
+
+#ifdef __APPLE__
+static void OSXMessageBox(std::string title, std::string message) {
```
const references?

lioncash commented on src/citra_qt/crash_handler.cpp:
```diff
@@ -0,0 +1,270 @@
+
+    std::string title = "Citra: Caught Unhandled Exception";
+
+    std::string message = std::string() +
```
```cpp
std::string message = std::string(
        "Press Ctrl+C to copy text\n"
        "Please also take a copy of the console window\n\n") +
        detail;
```

lioncash commented on src/citra_qt/crash_handler.cpp:
```diff
@@ -0,0 +1,270 @@
+
+        // Get source code line information.
+        DWORD line_displacement = 0; // The number of bytes return_address is offset from the first instruction of this line.
+        IMAGEHLP_LINE64 line = { 0 };
```
the zero here isn't necessary.

lioncash commented on src/citra_qt/crash_handler.cpp:
```diff
@@ -0,0 +1,270 @@
+ * @param ep The exception pointer containing exception information.
+ * @return See Microsoft's documentation on SetUnhandledExceptionFilter for possible return values.
+ */
+static LONG WINAPI MyUnhandledExceptionFilter(struct _EXCEPTION_POINTERS* ep) {
```
`struct` isn't necessary here.

lioncash commented on src/citra_qt/crash_handler.cpp:
```diff
@@ -0,0 +1,270 @@
+ * @param c The context record that contains the information on the stack of interest.
+ * @return A string containing a human-readable stack trace.
+ */
+static std::string GetStackTrace(CONTEXT& c) {
```
`ctx` is likely a better shorthand here.

lioncash commented on src/citra_qt/crash_handler.cpp:
```diff
@@ -0,0 +1,270 @@
+        }
+
+        // Format string
+        std::string ret;
```
Actually, this can just use StringFromFormat in string_util.h and just be:

```cpp
return Common::StringFromFormat("[%llx] %s+0x%llx (%s:%i)\n", return_address, undecorated_name, symbol_displacement, file_name.c_str(), line.LineNumber);
```

lioncash commented on src/citra_qt/crash_handler.cpp:
```diff
@@ -0,0 +1,270 @@
+        constexpr size_t SYMBOL_NAME_SIZE = 512; // arbitrary value
+
+        // Allocate space for symbol info.
+        IMAGEHLP_SYMBOL64* symbol = static_cast<IMAGEHLP_SYMBOL64*>(calloc(sizeof(IMAGEHLP_SYMBOL64) + SYMBOL_NAME_SIZE * sizeof(char), 1));
```
why `calloc`?

MerryMage replied:
`calloc` instead of `malloc` because I want to zero the memory. The Win32 API (still called Win32 on x64) often requires unused members to be set to zero.

Not stack-allocated because I don't think it's possible to do so.

lioncash replied:
Hmm, looked it up and yeah, it's MS being silly/quirky with its API (technically it can be done on the stack with an instance and a trailing buffer used for the name, but it's relying on memory locations, so it's not worth it).

yuriks replied:
This could be done with a struct:
```
struct {
  IMAGEHLP_SYMBOL64 info;
  char name[SYMBOL_NAME_SIZE];
} symbol;
```

lioncash commented on src/citra_qt/crash_handler.cpp:
```diff
@@ -0,0 +1,270 @@
+
+    std::string message = std::string() +
+        "Press Ctrl+C to copy text\n" +
+        "Please also take a copy of the console window\n" +
```
This might want to be phrased as taking a screenshot, for better clarity.

lioncash replied:
(it's somewhat unfortunate this can't be/isn't hooked by Qt, since this can't be localized in its current state)

MerryMage replied:
Translated strings could be passed in if they were arguments of `Register()`. Thoughts?

lioncash replied:
One way to do it is to possibly provide the ability for the UI side to register a callback with the crash handler of the form `void CrashHandlerCallback(const std::string& stack_trace)`, which would forward the retrieved stack trace to any arbitrary front-end, and allow for localizing dialog title/user instructions. However, considering code comments say we might not be on the GUI thread when it happens, I don't know how feasible that would be.

lioncash replied:
There's also [this Qt documentation](https://doc.qt.io/qt-5/unix-signals.html) and [this](https://github.com/JPNaude/dev_notes/wiki/Produce-a-stacktrace-when-something-goes-wrong-in-your-application) which may be useful.

Subv commented on src/citra_qt/crash_handler.cpp:
```diff
@@ -0,0 +1,270 @@
+static LONG WINAPI MyUnhandledExceptionFilter(struct _EXCEPTION_POINTERS* ep) {
+    std::string stack_trace = GetStackTrace(*ep->ContextRecord);
+
+    std::string detail = std::string() +
```
Why is the std::string() needed?

MerryMage replied:
Concatenation. I could do `std::string("Version: ")` instead if preferable.

MerryMage commented:
Addressed above comments except @lioncash's [comment on localisation](https://github.com/citra-emu/citra/pull/1874#discussion_r65587189).

Tested on OS X and added a screenshot.

lioncash commented on src/citra_qt/crash_handler.cpp:
```diff
@@ -0,0 +1,277 @@
+    if (num_frames == 0)
+        return "";
+
+    char** symbols = backtrace_symbols(return_addresses, num_frames);
```
This needs to be freed before the function ends.

lioncash commented on src/citra_qt/crash_handler.cpp:
```diff
@@ -0,0 +1,277 @@
+namespace CrashHandler {
+
+static LONG WINAPI MyUnhandledExceptionFilter(_EXCEPTION_POINTERS*);
+static std::string GetStackTrace(CONTEXT& c);
```
the `c` here can be omitted for consistency with the filter function above it.

lioncash commented on src/citra_qt/crash_handler.cpp:
```diff
@@ -0,0 +1,277 @@
+
+    std::string ret;
+    for (size_t i = 0; i < num_frames; i++) {
+        fprintf(stderr, "%s\n", symbols[i]);
```
This fprintf should likely be inside SignalHandler alongside the `fflush()` call, since it can just be one fprintf call after all the concatenation is done.

lioncash replied:
Actually, if you want to be signal-handler safe, technically you'd use POSIX `write` instead of fprintf to write to stderr.

wwylele commented:
Here is a test result from Ubuntu: (by intentionally trigger an ASSERT)
```
[  24.973505] Debug <Critical> core/memory.cpp:operator():146: Assertion Failed!
Unreachable code!
Oops. Now in CrashHandler::SignalHandler:
Caught signal 5 (SIGTRAP)
Git Branch: 1874 (fa25484)
Git Commit: fa25484515e2318684bc6aceae896271d9f86fb4
Stack Trace:
citra-qt() [0x4b7492]
/lib/x86_64-linux-gnu/libc.so.6(+0x354a0) [0x7f55e25184a0]
citra-qt() [0x4ec7d1]
citra-qt() [0x4eae3b]
citra-qt(_ZN21RasterizerCacheOpenGL11FlushRegionEjjPK13CachedSurfaceb+0x2b4) [0x55d624]
citra-qt() [0x4ea41b]
citra-qt(_ZN6Kernel9VMManager10UnmapRangeEjj+0x118) [0x4d6ed8]
citra-qt(_ZN6Kernel7Process10LinearFreeEjj+0xa6) [0x51af66]
citra-qt() [0x531a88]
citra-qt(_ZN3SVC7CallSVCEj+0x74) [0x5317e4]
citra-qt(_Z19InterpreterMainLoopP11ARMul_State+0xc566) [0x4fc526]
citra-qt(_ZTv0_n168_N10ARM_DynCom19ExecuteInstructionsEi+0x21) [0x4efb21]
citra-qt(_ZN4Core7RunLoopEi+0x51) [0x4cadf1]
citra-qt(_ZN9EmuThread3runEv+0x5f) [0x4b1eaf]
/usr/lib/x86_64-linux-gnu/libQt5Core.so.5(+0xa584e) [0x7f55e340f84e]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x76fa) [0x7f55e28b36fa]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7f55e25e9b5d]
QObject::~QObject: Timers cannot be stopped from another thread
QSocketNotifier: Invalid socket 28 and type 'Read', disabling...
```

lioncash commented:
Since Github doesn't un-collapse conversations with new comments, just leaving this here to indicate that I've added to my localization comment.

mailwl commented:
Unfortunally, doesn't help with Exstetra #1629 (both Debug/Release). IDE callstack doesn't help, too

![extetracrash](https://cloud.githubusercontent.com/assets/629437/15799246/8848ed44-2a5e-11e6-8cda-467bf6c04fd3.png)

JayFoxRox commented:
Isn't there a nice way to show a text box with a text field inside?

I don't like the design of the messagebox at all. It will scare users.
Also I think telling them to Ctrl+C is black magic for some users (*cough*pokemon crowd*cough*).

\- I'd even prefer just printing the backtrace to terminal, then tell the user to copy the terminal content.
That way we will at least get log + backtrace together AND we don't necessarily scare users with weird numbers and text.
If we had log *files* this would be even easier.

MerryMage commented:
@JayFoxRox 
I'm considering redesigning this to actually use a watchdog process like I mention in my first `TODO` comment. That way we could: (a) have a nice cross-platform Qt crash dialog, (b) also redirect stdout/stderr to that watchdog so it can save a log when necessary.

@mailwl 
Crashes within JITted code are an interesting case given that the RSP and the structure of the stack are not going to be as expected.

mailwl commented:
Ah, I see, generated code not belong to citra code, so, no way to link with citra code

Subv commented:
What's the status of this?

yuriks commented on src/citra_qt/crash_handler.cpp:
```diff
@@ -0,0 +1,277 @@
+
+#include "common/string_util.h"
+
+namespace CrashHandler {
```
Whole file seems like a good addition to common instead of citra-qt

yuriks commented:
ping?

MerryMage commented:
Not getting back to this soon, but would like to see/do similar UI work in the future.
